File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/model/MTGCard.java
package com.dbottillo.mtgsearchfree.model;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import com.dbottillo.mtgsearchfree.R;

import java.util.ArrayList;
import java.util.List;

public class MTGCard implements Comparable<MTGCard>, Parcelable {

    private int id;
    private String name;
    private String type;
    private List<String> types;
    private List<String> subTypes;
    private List<Integer> colors;
    private int cmc;
    private String rarity;
    private String power;
    private String toughness;
    private String manaCost;
    private String text;
    private boolean isAMultiColor;
    private boolean isALand;
    private boolean isAnArtifact;
    private int multiVerseId;
    private MTGSet set;

    private int quantity;
    private boolean sideboard = false;
    private String layout;
    private String number;

    private List<String> rulings;

    private List<String> names;
    private List<String> superTypes;
    private String artist;
    private String flavor;
    private int loyalty;
    private List<String> printings;
    private String originalText;

    public MTGCard() {
        this.colors = new ArrayList<>();
        this.types = new ArrayList<>();
        this.subTypes = new ArrayList<>();
        this.rulings = new ArrayList<>();
        this.names = new ArrayList<>();
        this.superTypes = new ArrayList<>();
        this.printings = new ArrayList<>();
        this.quantity = 1;
        this.multiVerseId = -1;
    }

    public MTGCard(int id) {
        this();
        this.id = id;
    }

    public MTGCard(int id, int multiVerseId) {
        this(id);
        this.multiVerseId = multiVerseId;
    }

    public MTGCard(Parcel in) {
        this();
        readFromParcel(in);
    }

    public void addColor(String color) {
        addColor(CardProperties.COLOR.getNumberFromString(color));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeString(name);
        dest.writeString(type);
        dest.writeStringList(types);
        dest.writeStringList(subTypes);
        dest.writeList(colors);
        dest.writeInt(cmc);
        dest.writeString(rarity);
        dest.writeString(power);
        dest.writeString(toughness);
        dest.writeString(manaCost);
        dest.writeString(text);
        dest.writeInt(isAMultiColor ? 1 : 0);
        dest.writeInt(isALand ? 1 : 0);
        dest.writeInt(isAnArtifact ? 1 : 0);
        dest.writeInt(multiVerseId);
        dest.writeParcelable(set, flags);
        dest.writeInt(quantity);
        dest.writeInt(sideboard ? 1 : 0);
        dest.writeStringList(rulings);
        dest.writeString(layout);
        dest.writeString(number);
        dest.writeStringList(names);
        dest.writeStringList(superTypes);
        dest.writeString(artist);
        dest.writeString(flavor);
        dest.writeInt(loyalty);
        dest.writeStringList(printings);
        dest.writeString(originalText);
    }

    private void readFromParcel(Parcel in) {
        id = in.readInt();
        name = in.readString();
        type = in.readString();
        in.readStringList(types);
        in.readStringList(subTypes);
        in.readList(colors, Integer.class.getClassLoader());
        cmc = in.readInt();
        rarity = in.readString();
        power = in.readString();
        toughness = in.readString();
        manaCost = in.readString();
        text = in.readString();
        isAMultiColor = in.readInt() == 1;
        isALand = in.readInt() == 1;
        isAnArtifact = in.readInt() == 1;
        multiVerseId = in.readInt();
        set = in.readParcelable(MTGSet.class.getClassLoader());
        quantity = in.readInt();
        sideboard = in.readInt() == 1;
        in.readStringList(rulings);
        layout = in.readString();
        number = in.readString();
        in.readStringList(names);
        in.readStringList(superTypes);
        artist = in.readString();
        flavor = in.readString();
        loyalty = in.readInt();
        in.readStringList(printings);
        originalText = in.readString();
    }

    public static final Parcelable.Creator<MTGCard> CREATOR = new Parcelable.Creator<MTGCard>() {
        @Override
        public MTGCard createFromParcel(Parcel source) {
            return new MTGCard(source);
        }

        @Override
        public MTGCard[] newArray(int size) {
            return new MTGCard[size];
        }
    };


    @VisibleForTesting
    public void setId(int id) {
        this.id = id;
    }

    public long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public void setCardName(String name) {
        this.name = name;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public List<String> getTypes() {
        return types;
    }

    public void addType(String type) {
        this.types.add(type);
    }

    public List<String> getSubTypes() {
        return subTypes;
    }

    public void addSubType(String subType) {
        this.subTypes.add(subType);
    }

    public List<Integer> getColors() {
        return colors;
    }

    public void addColor(int color) {
        this.colors.add(color);
    }

    public int getCmc() {
        return cmc;
    }

    public void setCmc(int cmc) {
        this.cmc = cmc;
    }

    public String getRarity() {
        return rarity;
    }

    public void setRarity(String rarity) {
        this.rarity = rarity;
    }

    public String getPower() {
        return power;
    }

    public void setPower(String power) {
        this.power = power;
    }

    public String getToughness() {
        return toughness;
    }

    public void setToughness(String toughness) {
        this.toughness = toughness;
    }

    public String getManaCost() {
        return manaCost;
    }

    public void setManaCost(String manaCost) {
        this.manaCost = manaCost;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public boolean isMultiColor() {
        return isAMultiColor;
    }

    public void setMultiColor(boolean isMultiColor) {
        this.isAMultiColor = isMultiColor;
    }

    public boolean isLand() {
        return isALand;
    }

    public void setAsALand(boolean isALand) {
        this.isALand = isALand;
    }

    public boolean isArtifact() {
        return isAnArtifact;
    }

    public void setAsArtifact(boolean isAnArtifact) {
        this.isAnArtifact = isAnArtifact;
    }

    public boolean isEldrazi() {
        return !isMultiColor() && !isLand() && !isArtifact() && getColors().size() == 0;
    }

    public int getMultiVerseId() {
        return multiVerseId;
    }

    public void setMultiVerseId(int multiVerseId) {
        this.multiVerseId = multiVerseId;
    }

    public void belongsTo(MTGSet set) {
        this.set = set;
    }

    public MTGSet getSet() {
        return set;
    }

    public int getQuantity() {
        return quantity;
    }

    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }

    public boolean isSideboard() {
        return sideboard;
    }

    public void setSideboard(boolean sideboard) {
        this.sideboard = sideboard;
    }

    public String getLayout() {
        return layout;
    }

    public void setLayout(String layout) {
        this.layout = layout;
    }

    public String getNumber() {
        return number;
    }

    public void setNumber(String number) {
        this.number = number;
    }

    public String toString() {
        return "MTGCard: [" + name + "," + multiVerseId + "," + colors + "," + rarity + "]";
    }

    public void addRuling(String rule) {
        this.rulings.add(rule);
    }

    public void setColors(List<Integer> colors) {
        this.colors = colors;
    }

    public List<String> getNames() {
        return names;
    }

    public List<String> getSuperTypes() {
        return superTypes;
    }

    public String getFlavor() {
        return flavor;
    }

    public int getLoyalty() {
        return loyalty;
    }

    public void setNames(List<String> names) {
        this.names = names;
    }

    public void setSuperTypes(List<String> superTypes) {
        this.superTypes = superTypes;
    }

    public void setFlavor(String flavor) {
        this.flavor = flavor;
    }

    public void setLoyalty(int loyalty) {
        this.loyalty = loyalty;
    }

    public String getArtist() {
        return artist;
    }

    public void setArtist(String artist) {
        this.artist = artist;
    }

    public List<String> getPrintings() {
        return printings;
    }

    public void setPrintings(List<String> printings) {
        this.printings = printings;
    }

    public boolean isCommon() {
        return rarity.equalsIgnoreCase(CardFilter.FILTER_COMMON);
    }

    public boolean isUncommon() {
        return rarity.equalsIgnoreCase(CardFilter.FILTER_UNCOMMON);
    }

    public boolean isRare() {
        return rarity.equalsIgnoreCase(CardFilter.FILTER_RARE);
    }

    public boolean isMythicRare() {
        return rarity.equalsIgnoreCase(CardFilter.FILTER_MYHTIC);
    }

    public String getImage() {
        if (number != null && number.length() > 0
                && !set.getCode().equalsIgnoreCase("MM3")
                && !set.getCode().equalsIgnoreCase("MPS")
                && !set.getCode().equalsIgnoreCase("DDS")
                && !set.getCode().equalsIgnoreCase("AKH")
                && !set.getCode().equalsIgnoreCase("MPS_AKH")) {
            return "http://magiccards.info/scans/en/" + set.getMagicCardsInfoCode() + "/" + number + ".jpg";
        }
        return getImageFromGatherer();
    }

    public String getImageFromGatherer() {
        if (getMultiVerseId() > 0) {
            return "http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=" + getMultiVerseId() + "&type=card";
        }
        return null;
    }

    public List<String> getRulings() {
        return rulings;
    }

    public String getOriginalText() {
        return originalText;
    }

    public void setOriginalText(String originalText) {
        this.originalText = originalText;
    }

    @Override
    public int compareTo(@NonNull MTGCard card) {
        if (isALand && card.isALand) {
            return 0;
        }
        if (!isALand && card.isALand) {
            return -1;
        }
        if (isALand) {
            return 1;
        }
        if (isAnArtifact && card.isAnArtifact) {
            return 0;
        }
        if (!isAnArtifact && card.isAnArtifact) {
            return -1;
        }
        if (isAnArtifact) {
            return 1;
        }
        if (isAMultiColor && card.isAMultiColor) {
            return 0;
        }
        if (!isAMultiColor && card.isAMultiColor) {
            return -1;
        }
        if (isAMultiColor) {
            return 1;
        }

        if (card.getSingleColor() == this.getSingleColor()) {
            return 0;
        }
        if (getSingleColor() < card.getSingleColor()) {
            return -1;
        }
        return 1;
    }

    @VisibleForTesting
    public int getSingleColor() {
        if (isAMultiColor || getColors().size() == 0) {
            return -1;
        }
        return getColors().get(0);
    }

    public int getMtgColor(Context context) {
        int mtgColor = context.getResources().getColor(R.color.mtg_other);
        if (isMultiColor()) {
            mtgColor = context.getResources().getColor(R.color.mtg_multi);
        } else if (getColors().contains(CardProperties.COLOR.WHITE.getValue())) {
            mtgColor = context.getResources().getColor(R.color.mtg_white);
        } else if (getColors().contains(CardProperties.COLOR.BLUE.getValue())) {
            mtgColor = context.getResources().getColor(R.color.mtg_blue);
        } else if (getColors().contains(CardProperties.COLOR.BLACK.getValue())) {
            mtgColor = context.getResources().getColor(R.color.mtg_black);
        } else if (getColors().contains(CardProperties.COLOR.RED.getValue())) {
            mtgColor = context.getResources().getColor(R.color.mtg_red);
        } else if (getColors().contains(CardProperties.COLOR.GREEN.getValue())) {
            mtgColor = context.getResources().getColor(R.color.mtg_green);
        }
        return mtgColor;
    }

    @SuppressWarnings({"SimplifiableIfStatement", "RedundantIfStatement"})
    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        MTGCard other = (MTGCard) o;
        /*for (Field field : MTGCard.class.getDeclaredFields()) {
            if (!Modifier.isStatic(field.getModifiers())) {
                try {
                    if (field.get(this) != null && !field.get(this).equals(field.get(other))) {
                        Log.e("CUSTOM", "error on: " + field.getName() + " with values: " + field.get(this) + " vs " + field.get(other));
                    }
                } catch (IllegalAccessException e) {
                    LOG.e("impossible to read value");
                    e.printStackTrace();
                }
            }
        }*/

        if (!equalOrNull(name, other.name)) {
            return false;
        }
        if (multiVerseId > 0 && other.multiVerseId > 0 && !(multiVerseId == other.multiVerseId)) {
            return false;
        }
        if (!equalOrNull(type, other.type)) {
            return false;
        }
        if (!types.equals(other.types)) {
            return false;
        }
        if (!(subTypes == null && other.subTypes == null || (subTypes != null && other.subTypes != null && subTypes.equals(other.subTypes)))) {
            return false;
        }
        if (!(colors == null && other.colors == null || (colors != null && other.colors != null && colors.equals(other.colors)))) {
            return false;
        }
        if (!(cmc == other.cmc)) {
            return false;
        }
        if (!equalOrNull(rarity, other.rarity)) {
            return false;
        }
        if (!equalOrNull(power, other.power)) {
            return false;
        }
        if (!equalOrNull(toughness, other.toughness)) {
            return false;
        }
        if (!equalOrNull(manaCost, other.manaCost)) {
            return false;
        }
        if (!equalOrNull(text, other.text)) {
            return false;
        }
        if (!(isAMultiColor == other.isAMultiColor)) {
            return false;
        }
        if (!(isALand == other.isALand)) {
            return false;
        }
        if (!(set == null && other.set == null || (set != null && other.set != null && set.equals(other.set)))) {
            return false;
        }
        if (!equalOrNull(layout, other.layout)) {
            return false;
        }
        if (!equalOrNull(number, other.number)) {
            return false;
        }
        if (!(rulings == null && other.rulings == null || (rulings != null && other.rulings != null && rulings.equals(other.rulings)))) {
            return false;
        }
        if (!(names == null && other.names == null || (names != null && other.names != null && names.equals(other.names)))) {
            return false;
        }
        if (!(superTypes == null && other.superTypes == null || (superTypes != null && other.superTypes != null && superTypes.equals(other.superTypes)))) {
            return false;
        }
        if (!equalOrNull(flavor, other.flavor)) {
            return false;
        }
        if (!equalOrNull(artist, other.artist)) {
            return false;
        }
        if (!(loyalty == other.loyalty)) {
            return false;
        }
        if (!(printings == null && other.printings == null || (printings != null && other.printings != null && printings.equals(other.printings)))) {
            return false;
        }
        return true;
    }

    private boolean equalOrNull(String first, String second) {
        return (first == null && second == null || (first != null && second != null
                && first.equals(second)));
    }

    @Override
    public int hashCode() {
        return multiVerseId;
    }

    public boolean isWhite() {
        return manaCost != null && manaCost.contains("W");
    }

    public boolean isBlue() {
        return manaCost != null && manaCost.contains("U");
    }

    public boolean isBlack() {
        return manaCost != null && manaCost.contains("B");
    }

    public boolean isRed() {
        return manaCost != null && manaCost.contains("R");
    }

    public boolean isGreen() {
        return manaCost != null && manaCost.contains("G");
    }

    public boolean hasNoColor() {
        return manaCost == null || !manaCost.matches(".*[WUBRG].*");
    }

    public boolean isDoubleFaced() {
        return layout != null && layout.equalsIgnoreCase("double-faced");
    }

    public String bigToString() {
        return "MTGCard{"
                + "id=" + id
                + ", name='" + name + '\''
                + ", type='" + type + '\''
                + ", types=" + types
                + ", subTypes=" + subTypes
                + ", colors=" + colors
                + ", cmc=" + cmc
                + ", rarity='" + rarity + '\''
                + ", power='" + power + '\''
                + ", toughness='" + toughness + '\''
                + ", manaCost='" + manaCost + '\''
                + ", text='" + text + '\''
                + ", isAMultiColor=" + isAMultiColor
                + ", isALand=" + isALand
                + ", isAnArtifact=" + isAnArtifact
                + ", multiVerseId=" + multiVerseId
                + ", set=" + set
                + ", quantity=" + quantity
                + ", sideboard=" + sideboard
                + ", layout='" + layout + '\''
                + ", number='" + number + '\''
                + ", rulings=" + rulings
                + ", names=" + names
                + ", superTypes=" + superTypes
                + ", artist='" + artist + '\''
                + ", flavor='" + flavor + '\''
                + ", loyalty=" + loyalty
                + ", printings=" + printings
                + ", originalText=" + originalText
                + '}';
    }

}

File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/model/MTGCard.kt
package com.dbottillo.mtgsearchfree.model

import android.content.Context
import android.support.annotation.VisibleForTesting
import android.support.v4.content.ContextCompat
import com.dbottillo.mtgsearchfree.R

data class MTGCard(var id: Int = 0,
                   var name: String = "",
                   var type: String = "",
                   val types: MutableList<String> = mutableListOf(),
                   val subTypes: MutableList<String> = mutableListOf(),
                   var colors: MutableList<Int> = mutableListOf(),
                   var cmc: Int = 0,
                   var rarity: String = "",
                   var power: String = "",
                   var toughness: String = "",
                   var manaCost: String = "",
                   var text: String = "",
                   var isMultiColor: Boolean = false,
                   var isLand: Boolean = false,
                   var isArtifact: Boolean = false,
                   var multiVerseId: Int = 0,
                   var set: MTGSet? = null,
                   var quantity: Int = 1,
                   var isSideboard: Boolean = false,
                   var layout: String = "normal",
                   var number: String? = null,
                   val rulings: MutableList<String> = mutableListOf(),
                   var names: List<String> = listOf(),
                   var superTypes: List<String> = listOf(),
                   var artist: String = "",
                   var flavor: String? = null,
                   var loyalty: Int = 0,
                   var printings: List<String> = listOf(),
                   var originalText: String = "",
                   var mciNumber: String? = null,
                   var colorsIdentity: List<String>? = null,
                   var legalities: MutableList<Legality> = mutableListOf()) : Comparable<MTGCard> {

    constructor(onlyId: Int) : this(id = onlyId)

    constructor(onlyId: Int, multiVerseId: Int) : this(id = onlyId, multiVerseId = multiVerseId)

    fun setCardName(name: String) {
        this.name = name
    }

    fun addType(type: String) {
        this.types.add(type)
    }

    fun addSubType(subType: String) {
        this.subTypes.add(subType)
    }

    fun addColor(color: String) {
        colors.add(CardProperties.COLOR.getNumberFromString(color))
    }

    val isEldrazi: Boolean
        get() = !isMultiColor && !isLand && !isArtifact && colors.size == 0

    fun belongsTo(set: MTGSet) {
        this.set = set
    }

    override fun toString(): String {
        return "MTGCard: [$id,$name,$multiVerseId,$colors,$rarity,$quantity]"
    }

    fun addRuling(rule: String) {
        this.rulings.add(rule)
    }

    fun addLegality(legality: Legality) {
        legalities.add(legality)
    }

    val isCommon: Boolean
        get() = rarity.equals(CardFilter.FILTER_COMMON, ignoreCase = true)

    val isUncommon: Boolean
        get() = rarity.equals(CardFilter.FILTER_UNCOMMON, ignoreCase = true)

    val isRare: Boolean
        get() = rarity.equals(CardFilter.FILTER_RARE, ignoreCase = true)

    val isMythicRare: Boolean
        get() = rarity.equals(CardFilter.FILTER_MYHTIC, ignoreCase = true)

    val image: String?
        get() = if (number != null && set != null && number!!.isNotEmpty()
                && !types.contains("Plane")
                && set?.code?.toUpperCase() != "6ED") {
            "https://magiccards.info/scans/en/" + set?.magicCardsInfoCode + "/" + mciNumberOrMultiverseId + ".jpg"
        } else imageFromGatherer

    val imageFromGatherer: String?
        get() = if (multiVerseId > 0) {
            "http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=$multiVerseId&type=card"
        } else null

    private val mciNumberOrMultiverseId: String?
        get() = if (mciNumber.isNullOrEmpty()) {
            number
        } else mciNumber

    override fun compareTo(other: MTGCard): Int {
        if (isLand && other.isLand) {
            return 0
        }
        if (!isLand && other.isLand) {
            return -1
        }
        if (isLand) {
            return 1
        }
        if (isColorlessArtifact && other.isColorlessArtifact) {
            return 0
        }
        if (!isColorlessArtifact && other.isColorlessArtifact) {
            return -1
        }
        if (isColorlessArtifact) {
            return 1
        }
        if (isMultiColor && other.isMultiColor) {
            return 0
        }
        if (!isMultiColor && other.isMultiColor) {
            return -1
        }
        if (isMultiColor) {
            return 1
        }
        if (other.singleColor == this.singleColor) {
            return 0
        }
        return if (singleColor < other.singleColor) {
            -1
        } else {
            1
        }
    }

    private val isColorlessArtifact: Boolean
        get() = colorsIdentity?.let { it.isEmpty() && isArtifact } ?: isArtifact

    val singleColor: Int
        @VisibleForTesting
        get() =
            when {
                isMultiColor -> -1
                colorsIdentity != null -> colorsIdentity!![0].toColorInt()
                colors.isNotEmpty() -> colors[0]
                else -> -1
            }

    fun getMtgColor(context: Context): Int {
        return ContextCompat.getColor(context,
                when {
                    isMultiColor -> R.color.mtg_multi
                    colors.contains(CardProperties.COLOR.WHITE.value) -> R.color.mtg_white
                    colors.contains(CardProperties.COLOR.BLUE.value) -> R.color.mtg_blue
                    colors.contains(CardProperties.COLOR.BLACK.value) -> R.color.mtg_black
                    colors.contains(CardProperties.COLOR.RED.value) -> R.color.mtg_red
                    colors.contains(CardProperties.COLOR.GREEN.value) -> R.color.mtg_green
                    else -> R.color.mtg_other
                })
    }

    val isWhite: Boolean
        get() = manaCost.contains("W") || colorsIdentity?.contains("W") ?: false

    val isBlue: Boolean
        get() = manaCost.contains("U") || colorsIdentity?.contains("U") ?: false

    val isBlack: Boolean
        get() = manaCost.contains("B") || colorsIdentity?.contains("B") ?: false

    val isRed: Boolean
        get() = manaCost.contains("R") || colorsIdentity?.contains("R") ?: false

    val isGreen: Boolean
        get() = manaCost.contains("G") || colorsIdentity?.contains("G") ?: false

    fun hasNoColor(): Boolean {
        return !manaCost.matches(".*[WUBRG].*".toRegex()) && colorsIdentity?.isEmpty() ?: false
    }

    val isDoubleFaced: Boolean
        get() = layout.equals("double-faced", ignoreCase = true)

    override fun equals(other: Any?): Boolean {
        return when {
            other !is MTGCard -> false
            multiVerseId > 0 && other.multiVerseId == multiVerseId -> true
            name.isNotEmpty() && other.name == name -> true
            else -> false
        }
    }

}

class Legality(val format: String, val legality: String)

fun String.toColorInt(): Int {
    return when (this) {
        "W" -> 0
        "U" -> 1
        "B" -> 2
        "R" -> 3
        "G" -> 4
        else -> 1
    }
}